cdc/changefeed.go

```diff
@@ -712,6 +724,33 @@ func (c *changeFeed) calcResolvedTs(ctx context.Context) error {
 		}
 	}
 
+	//sync-point on
+	if c.info.SyncPointEnabled {
+		c.syncpointMutex.Lock()
+		defer c.syncpointMutex.Unlock()
+		if c.status.ResolvedTs == c.status.CheckpointTs && !c.updateResolvedTs {
+			log.Info("sync point reached by ticker", zap.Uint64("ResolvedTs", c.status.ResolvedTs), zap.Uint64("CheckpointTs", c.status.CheckpointTs), zap.Bool("updateResolvedTs", c.updateResolvedTs), zap.Uint64("ddlResolvedTs", c.ddlResolvedTs), zap.Uint64("ddlTs", c.ddlTs))
+			c.updateResolvedTs = true
+			err := c.sinkSyncpoint(ctx)
+			if err != nil {
+				log.Error("syncpoint sink fail", zap.Uint64("ResolvedTs", c.status.ResolvedTs), zap.Uint64("CheckpointTs", c.status.CheckpointTs), zap.Error(err))
+			}
+		}
```

Comment:
CI is not stable, ref: https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/cdc_ghpr_integration_test/detail/cdc_ghpr_integration_test/2679/pipeline/

The following check fails

```
[2020-09-29T05:02:45.506Z] time="2020-09-29T13:02:43+08:00" level=info msg="set history snapshot: SET @@tidb_snapshot='419785771227283458'"
[2020-09-29T05:02:45.506Z] time="2020-09-29T13:02:43+08:00" level=info msg="set history snapshot: SET @@tidb_snapshot='419785771294392321'"
[2020-09-29T05:02:45.506Z] check result: fail!
[2020-09-29T05:02:45.506Z] 2 tables' check passed, 1 tables' check failed.
[2020-09-29T05:02:45.506Z] 
[2020-09-29T05:02:45.506Z] table: testSync.simple1
[2020-09-29T05:02:45.506Z] table's struct not equal
[2020-09-29T05:02:45.506Z] table's data equal
```

This `syncpoint` is recorded by ticker, the checkpoint ts reaches DDL FinishedTs, but the DDL has not been executed in downstream.

```
[2020-09-29T05:03:05.759Z] [2020/09/29 13:00:33.916 +08:00] [INFO] [changefeed.go:732] ["sync point reached by ticker"] [ResolvedTs=419785771227283458] [CheckpointTs=419785771227283458] [updateResolvedTs=false] [ddlResolvedTs=419785771227283458] [ddlTs=0]
```

Reply:
```suggestion
		if c.status.ResolvedTs == c.status.CheckpointTs && !c.updateResolvedTs && c.ddlTs==0 {
			log.Info("sync point reached by ticker", zap.Uint64("ResolvedTs", c.status.ResolvedTs), zap.Uint64("CheckpointTs", c.status.CheckpointTs), zap.Bool("updateResolvedTs", c.updateResolvedTs), zap.Uint64("ddlResolvedTs", c.ddlResolvedTs), zap.Uint64("ddlTs", c.ddlTs))
			c.updateResolvedTs = true
			err := c.sinkSyncpoint(ctx)
			if err != nil {
				log.Error("syncpoint sink fail", zap.Uint64("ResolvedTs", c.status.ResolvedTs), zap.Uint64("CheckpointTs", c.status.CheckpointTs), zap.Error(err))
			}
		}
```
I think add the condition about c.ddlTs can fix this bug, if c.ddlTs!=0 that means we must wait a DDL to sink and the syncpoint by ticker will be  ignore,if the c.ddlTs==0 that means there no DDL and we should sink syncpoint by ticker.

Reply:
emmm, if we use `c.ddlTs`, should we move these code after `c.ddlTs` updated (L835).
In the log, `c.ddlTs == 0`

```
[2020-09-29T05:03:05.759Z] [2020/09/29 13:00:33.916 +08:00] [INFO] [changefeed.go:732] ["sync point reached by ticker"] [ResolvedTs=419785771227283458] [CheckpointTs=419785771227283458] [updateResolvedTs=false] [ddlResolvedTs=419785771227283458] [ddlTs=0]
```

Reply:
Yes, you are right! We should change the order order of the code.